samples/volumes/example5.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import kfp.dsl as dsl
+
+
+@dsl.pipeline(
+    name="Example 5",
+    description="The fifth example of the design doc."
+)
+def example5(rok_url):
+    vol1 = dsl.PipelineVolume(
+        name="vol1",
+        size="1Gi",
+        annotations={"rok/origin": rok_url},
+        mode=dsl.VOLUME_MODE_RWM
+    )
+
+    step1 = dsl.ContainerOp(
+        name="step1_concat",
+        image="library/bash:4.4.23",
+        command=["sh", "-c"],
+        arguments=["cat /data/file*| gzip -c >/data/full.gz"],
+        volumes={"/data": vol1}
+    )
```

Comment:
If we change vol1 as an output of a PvcOp, we can embed the output PipelineParam of PvcOp in the volume spec. In the containerop, we can leverage the similar way as args and command to parse the volumes to get input params. In this way, PipelineVolume can be served as a pure data object and we can just rely on PipelineParam to manage the data passing and dependency.

Reply:
@hongye-sun
> If we change vol1 as an output of a PvcOp, we can embed the output PipelineParam of PvcOp in the volume spec. In the containerop, we can leverage the similar way as args and command to parse the volumes to get input params. In this way, PipelineVolume can be served as a pure data object and we can just rely on PipelineParam to manage the data passing and dependency.

Yes, this works. But has two disadvantanges, let's continue the discussion in #801:

1. It requires modifications to Argo to allow `container` templates to consume `volumes` that contain input parameters that must be evaluated before each container task in the DAG
2. It cannot capture the case when there is multiple container tasks are using the same volume and we need to depend on *all* containers finishing, before another task can consume [mount] the volume.